modelkit/core/profilers/simple.py

```diff
@@ -105,26 +119,61 @@ def summary(  # type: ignore
             return tabulate(result_sorted, headers="keys", **kwargs)
         return result_sorted
 
+    def _get_current_model_calls(self, model_name: str) -> Dict[str, int]:
+        """Get the number of current sub model calls.
+        Args:
+            model_name (str)
+        Returns:
+            Dict[str, int]: sub model calls
+        """
+        current_calls: Dict[str, int] = {}
+        for sub_model in self.graph[model_name]:
+            current_calls[sub_model] = self.graph_calls[sub_model]["__main__"]
+        return current_calls
+
+    def _compute_direct_calls_and_update_graph_calls(
+        self, model_name: str, previous_calls: Dict[str, int]
+    ) -> Dict[str, int]:
+        self.graph_calls[model_name]["__main__"] += 1
+        direct_calls: Dict[str, int] = {}
+        for sub_model in self.graph[model_name]:
+            direct_calls[sub_model] = (
+                self.graph_calls[sub_model]["__main__"] - previous_calls[sub_model]
+            )
+            self.graph_calls[model_name][sub_model] += direct_calls[sub_model]
+        # remove non-direct calls
+        for sub_model in direct_calls:
+            for sub_sub in self.graph_calls[sub_model]:
+                if direct_calls[sub_model] > 0 and (sub_sub in direct_calls):
+                    direct_calls[sub_sub] -= self._get_avg_call(sub_model, sub_sub)
+        return direct_calls
+
+    def _get_avg_call(self, model: str, sub_model: str) -> int:
+        """Compute average calls of sub_model in model
+        Returns:
+            int: average calls of sub_model in model
+        """
+        return int(
+            self.graph_calls[model][sub_model] / self.graph_calls[model]["__main__"]
+        )
+
     def _calculate_net_cost(
-        self, model_names: List[str], model_costs: List[float], graph: Dict[str, Set]
-    ) -> List[float]:
-        """Compute net cost of each sub models.
+        self, duration: float, direct_calls: Dict[str, int]
+    ) -> float:
+        """Compute net cost of each sub models. Subtracting model "duration" by
+        "duration" of all direct sub model.
+
         Args:
-            model_names (List[str]): list of model name
-            model_costs (List[float]): cost of each model
-            graph (Dict[str, Set]): graph[model_name] = set of direct sub model names
+            duration (float): model duration
+            direct_calls (Dict[str, int]): direct calls of sub_model
 
         Returns:
-            List[float]: net cost
+            float: net cost
         """
-        result = []
-        data = dict(zip(model_names, model_costs))
-        for model_name in model_names:
-            result.append(
-                max(
-                    data[model_name]
-                    - sum(data.get(sub_model, 0) for sub_model in graph[model_name]),
-                    0,
+        net_duration = duration
+        for sub_model in direct_calls:
```

Comment:
NB:
```
        for sub_model, num_calls in direct_calls.items():
            if num_calls > 0:
                net_duration -= sum(self.durations[sub_model][-num_calls:])
```

Reply:
Thank you for the suggestion, the code has been updated.